### ISSUE TITLE
FEProblemBase::getMaxQps() becomes invalid after repartitioning or mesh changes.

### ISSUE THREAD

**Issue by roystgnr** [idaholab/moose#8462]
### Description of the enhancement or error report

FEProblemBase calculates _max_qps (and _max_shape_funcs) once, on the active local elements of the initial mesh.  This value gets queried for critical and permanent uses such as MaterialProperty array sizes.

The value then becomes outdated if/when a processor is given elements, e.g. via repartitioning, that require more quadrature points than were originally expected.  opt-mode segfaults or devel/dbg-mode assertion failures ensue.

### Rationale for the enhancement or information for reproducing the error

This is easily reproduced via
```./run_tests -p 5 --re controls/time_periods/dgkernels.test```

In theory "running with more processors than coarse elements" isn't the only possible failure case, however; any hybrid mesh run in parallel is a failure risk.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

I'll have a small PR shortly which fixes the repartitioning-triggered issue.

Other ways of adding quadrature points, such as adaptive p refinement, will require a different and probably much more difficult fix.  If I ever convince @friedmud that p adaptivity is awesome and not WONT_FIX then I'll create a new issue for that.
